docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,841 @@
+    worker: {
+      dev: {
+        createEnvironment: createWorkerEnvironment,
+      },
+    },
+  },
+})
+```
+
+:::
+
+`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together. If you don't use either of them, then you can define your own function to communicate between the runner and the server.
+
+## ModuleRunnerHMRConnection
+
+**Type Signature:**
+
+```ts
+export interface ModuleRunnerHMRConnection {
+  /**
+   * Checked before sending messages to the client.
+   */
+  isReady(): boolean
+  /**
+   * Send a message to the client.
+   */
+  send(message: string): void
+  /**
+   * Configure how HMR is handled when this connection triggers an update.
+   * This method expects that the connection will start listening for HMR updates and call this callback when it's received.
+   */
+  onUpdate(callback: (payload: HMRPayload) => void): void
+}
+```
+
+This interface defines how HMR communication is established. Vite exports `ServerHMRConnector` from the main entry point to support HMR during Vite SSR. The `isReady` and `send` methods are usually called when the custom event is triggered (like, `import.meta.hot.send("my-event")`).
+
+`onUpdate` is called only once when the new module runner is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:
+
+```js
+function onUpdate(callback) {
+  this.connection.on('hmr', (event) => callback(event.data))
+}
+```
+
+The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in a module runner will wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.
+
+## Environments during build
+
+Plugin hooks also receive the environment instance during build. This replaces the `ssr` boolean we have been passing them so far.
+
+In the CLI, calling `vite build` will build the Client. It is equivalent to calling `vite build --environment=client`.
+
+The build the SSR server, the `--ssr` shourcut can be used: `vite build --ssr`. This is equivalent to calling `vite build --environment=ssr`.
+
+Other non-default environments can be build using `vite build --environment=name`.
+
+## Building all environments
+
+Calling `vite build --all` will instantiate a `ViteBuilder` (build-time equivalent to a `ViteDevServer`) to build all configured environments for production. By default the build of environments is run in series respecting the order of the `environments` array. A framework or user can further configure how the build tasks are excecuted using:
+
+```js
+export default {
+  builder: {
+    runBuildTasks: asnyc (builder, buildTasks) => {
+      return Promise.all(buildTasks.map( task => task.run() ))
```

Comment:
Right now, the implementation resolves the config for each build generating a separate pipeline. I did it this way so we can parallelize without issues given the current state of plugins in the ecosystem.

@antfu proposed that we should take the opportunity and try to align with dev, and have a single plugins pipeline during build so it is easier to share information during build time too. We're going to push everyone to use something like `this.environment.cache` if they need to store things while processing (maybe we need a `this.environment.api` for comms?), if the plugin isn't environment agnostic. In dev, this isn't a stretch as people are already used to do the same with the `ssr` flag (probably there are a lot of bugs due to improper isolation because things aren't as clear as having a proper environment abstraction). During build, we could do the same and ask everyone to be aware of the environment that is being built. Running the plugins in parallel, even for workers (that we may rethink as environments too) becomes possible.

I'm worried about this idea though. I think it is an interesting goal to aim for, but a lot of plugins will be broken for `vite build --all` if we do this. I was proposing that we have a `builder.isolatedPlugins` flag that we make opt-in for a major or two and let users set it if all their plugins work well in this new world.

Reply:
About `buildStart`, we actually would like to go the other way around and call `buildStart` during dev once per environment and make plugins use that to reset the state of each environment. This could be very interesting if at one point we would like to implement a "restart this environment" if a config changes for example without the need for a full server restart.

Reply:
Is there a way for plugin to know the start or end of certain environment build?
For example, currently it doesn't look like `environment` is accessible from `buildStart`, `buildEnd`, `writeBundle` etc...

Reply:
We need to implement that, `this.environment` should be available for all hooks. So far we only injected it in `resolveId`, `load`, and `transform` as it was done before.
I wonder if Rollup could give us a better way to inject the environment so we don't need to wrap every hook, that feels quite inefficient